oqupy/tempo.py

```diff
@@ -90,7 +92,7 @@ class TempoParameters(BaseAPIClass):
     def __init__(
             self,
             dt: float,
-            dkmax: int,
+            tcut: Union[float, int, None],
             epsrel: float,
```

Comment:
I guess that this might be a little bit of a trap for some users that might expect that tcut=1.9, tcut=2, tcut=2.1 behave similarly.
I suggest to either drop dkmax completely or have both optionally. In fact having both would ensure some backwards compatibility. There shouldn't be a problem with all four combinations of setting/not setting tcut and dkmax. In particular there is no real problem in case somone specifies both (tcut and dkmax) because one could simply then choose the effective minimum. Therefore I suggest this header:
```python
    def __init__(
            self,
            dt: float,
            dkmax: int,
            epsrel: float,
            tcut: Optional[float] = None,
            dkmax: Optional[int] = None,
            add_correlation_time: Optional[float] = None,
            subdiv_limit: Optional[int] = SUBDIV_LIMIT,
            liouvillian_epsrel: Optional[float] = INTEGRATE_EPSREL,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
```

Reply:
Hi @gefux,

Yes I would prefer to have both optional.

I also (quite strongly) don't like the type-dependent behaviour but decided to go that way based on the code in `contractions.compute_correlations` (implementing mutually exclusive parameters could be viewed as un-pythonic?). Hence I think we should review how we do things there too. 

I'll need to rewrite the input parsing and setters etc. so will update the PR when that's done.

Reply:
Another aspect that personally caused me difficultly (as a developer) is the interaction between getteres, setters and 'private' (i.e. underscore) variables in some of the classes. It seems quite confusion and inconsistent in places. I think I understand how it works now (it looks like __init__ doesn't set the underscore variables, but it is because of the getter methods), but wanted to ask if there is a particular style guide/paradigm you have been following when writing that structure? Would be helpful if I had a reference.

Reply:
ad type dependence: the code in `contractions.comput_correlations` might indeed not be ideal, however, I didn't have a much better idea so far. 
Here however, it is a bit different because I don't see a problem with setting both `tcut` and `dkmax`, which means that they are not necessarily mutually exclusive. If you cut twice you are left with the minimum of both cuts.

ad setter and getter methods: I agree that things are not ideal (in particular the parsing is all over the place). Getter and setter functions are just a more versatile way of having public object variables (such as `dt` in `TempoParameters`). My reasoning has mostly been:

- If a public object variable must not be changed after the object construction then there should be a getter function but no setter function, and the value (stored in the corresponding private variable) must be set in the constructor. This ensures that an error is thrown when the user tries to set that variable. For example: if `system` is a `System` object, then `system.dimension` is the Hilbert space dimension but `system.dimension = 42` throws an error.
- If a public object variable may be changed after  the object construction then there using getter, setter, deleter gives some extra space to do parsing and define special behaviour. For example (looking into `BaseAPIClass`)  you can set `system.name="3-level-sys"` which will check if the `"3-level-sys"` is a text, and if you delete it with `del system.name` it will in fact set it to `"__unnamed__"`.

I have no doubt that all of this could be done in a cleaner / more consistant way. Please share in an Issue if you have concrete suggestions!

Reply:
Thanks for explaining that Gerald, yes I can see the versatility they bring, makes a very neat interface and I'm not suggesting we change that (but I'll let you know if I ever do come across a different/better approach). I do also think we should bear in mind barrier to entry/complexity newcommers who might make contributions to the codebase, since we want to encourage development! Hence thought I'd mention that I personally found that tricky to read/parse when I wanted to make a change.

In the current implementation I've prevented `dkmax` and `tcut` from being set simultaneously in `__init__`, because I think that could easily lead to error where someone accidentally leaves one of the arguments in without realising, or doesn't read a docstring explaining that the minimum would be taken etc. If you feel strongly they should be able to be passed at the same time, let me know.

Reply:
Great, happy to make the required changes there. By the same reasoning we probably don't want deleters (except for name/description), would you agree?

Reply:
Okay if you don't have @deleter then trying to delete the attribute raises an AttributeError. I think that's a reasonable behaviour.

Reply:
Proposed changes made now [d6e8776]

When we've finalised things I should do a rebase.